### PR TITLE
[llvm] Drop code for llvm 15.

### PR DESCRIPTION
**Pull request by python3kgae**
Still use clang10 for COMPILE_LLVM_RUNTIME.
Most changes are related to the opaque ptr change in llvm 15.
It requires type when create Load and GEP. It would be nice to generate the type from taichi type.

Known issue.
Still some getPointerElementType use which is Deprecated for llvm 15.
Also some cases use hack to get llvm type for Load/GEP.

Only tested python tests/run_tests.py -v -t3 -a cpu -s on windows.
2 crashes.
crashes:
  fp16 crash for my hw not support fp16 for vulkan.
  element_wise crashed in llvm slp-vectorizer pass on skylake cpu, Already fixed upstream llvm.

Also tested clang15 for COMPILE_LLVM_RUNTIME, same result.

Related issue = #5276

### CODE REVIEW

**Comment by k-ye** [comment]
This is gonna be soooo cool! If I recall correctly, there are three parts involving LLVM:

1. We still use `clang-10` for compiling https://github.com/taichi-dev/taichi/blob/master/taichi/runtime/llvm/runtime_module/runtime.cpp into an LLVM module (i.e. `compile_llvm_runtime`).
2. We use LLVM-15 library for Taichi's LLVM backends
3. IIRC, there were errors when compiling Taichi codebase itself with `>= clang-12`.

This could be a great opportunity for us to unify the toolchain, upgrade them to a consistent version of LLVM-15, and consolidate our docs for how to upgrade LLVM in Taichi...

So far the biggest concern is still the slow LLVM toolchain installation time (https://github.com/taichi-dev/taichi/issues/5276#issuecomment-1168437523). Any other foreseeable difficulties in the upgrade? @ailzhang @bobcao3 @qiao-bo

**Review by qiao-bo — APPROVED** [review]
Thanks, looks good to me. We better have an issue to track the progress and aftermath when this is merged.

**Comment by jim19930609** [comment]
Hi @python3kgae, thanks so much for this awesome contribution! 

I'd also love to take a look at this PR and was wondering if you can keep this PR opened for a while? Thanks!

**Comment by python3kgae** [comment]
> Hi @python3kgae, thanks so much for this awesome contribution!
> 
> I'd also love to take a look at this PR and was wondering if you can keep this PR opened for a while? Thanks!

Sure.
It would be nice to have some advice for how to get llvm type from taichi type. Current code to get llvm type is hacky.

File: taichi/codegen/llvm/codegen_llvm.cpp [hunk 1]
```diff
@@ -1492,20 +1560,49 @@ llvm::Value *CodeGenLLVM::create_bit_ptr(llvm::Value *byte_ptr,
...
   return bit_ptr;
 }
 
 std::tuple<llvm::Value *, llvm::Value *> CodeGenLLVM::load_bit_ptr(
     llvm::Value *bit_ptr) {
-  auto byte_ptr = builder->CreateLoad(builder->CreateGEP(
-      bit_ptr, {tlctx->get_constant(0), tlctx->get_constant(0)}));
-  auto bit_offset = builder->CreateLoad(builder->CreateGEP(
-      bit_ptr, {tlctx->get_constant(0), tlctx->get_constant(1)}));
+#ifdef TI_LLVM_15
```
Comment (jim19930609) [inline taichi/codegen/llvm/codegen_llvm.cpp:1580]: Looks like there are too many TI_LLVM_15s here. Can we separate them into two code paths, and simply write:
```
#ifdef TI_LLVM_15
....
#else
....
#endif
```

Just for code cleanliness :)

File: taichi/codegen/llvm/codegen_llvm.cpp [hunk 2]
```diff
@@ -1110,7 +1141,22 @@ void CodeGenLLVM::visit(ReturnStmt *stmt) {
...
+  auto *val = llvm_val[stmt->src[0].var];
+  if (auto *alloc = llvm::dyn_cast<llvm::AllocaInst>(val))
+    ptr_ty = alloc->getAllocatedType();
+  if (!ptr_ty && stmt->src[0].var->element_type().is_pointer()) {
+    ptr_ty = llvm_type(stmt->src[0].var->element_type().ptr_removed());
+  }
+  TI_ASSERT(ptr_ty);
+#endif
+  llvm_val[stmt] = builder->CreateLoad(
+#ifdef TI_LLVM_15
```
Comment (jim19930609) [inline taichi/codegen/llvm/codegen_llvm.cpp:1156]: We probably wanna reduce the use of TI_LLVM_15 macro in the same scope. Can we join these two TI_LLVM_15 macros?

File: taichi/codegen/llvm/codegen_llvm.cpp [hunk 3]
```diff
@@ -1523,7 +1620,20 @@ void CodeGenLLVM::visit(SNodeLookupStmt *stmt) {
...
+    // FIXME: get parent_type from taichi instead of llvm.
+    llvm::Type *parent_type = builder->getInt8Ty();
+    if (auto bit_cast = llvm::dyn_cast<llvm::BitCastInst>(parent)) {
+      parent_type = bit_cast->getDestTy();
+      if (auto ptr_ty = llvm::dyn_cast<llvm::PointerType>(parent_type))
+        parent_type = ptr_ty->getPointerElementType();
+    }
+#endif
+    llvm_val[stmt] = builder->CreateGEP(
+#ifdef TI_LLVM_15
```
Comment (jim19930609) [inline taichi/codegen/llvm/codegen_llvm.cpp:1633]: ditto.

File: taichi/codegen/llvm/codegen_llvm.cpp [hunk 4]
```diff
@@ -1571,8 +1681,30 @@ void CodeGenLLVM::visit(GetChStmt *stmt) {
...
+        ptr_ty = tlctx->get_data_type(
+            tmpo_stmt->ret_type->cast<TensorType>()->get_element_type());
+      } else {
+        ptr_ty = tlctx->get_data_type(tmpo_stmt->ret_type.ptr_removed());
+      }
+    }
+    TI_ASSERT(ptr_ty);
+#endif
+    llvm_val[stmt] = builder->CreateGEP(
+#ifdef TI_LLVM_15
```
Comment (jim19930609) [inline taichi/codegen/llvm/codegen_llvm.cpp:1704]: ditto.

File: taichi/codegen/llvm/codegen_llvm.cpp [hunk 5]
```diff
@@ -2038,10 +2219,31 @@ void CodeGenLLVM::visit(BlockCornerIndexStmt *stmt) {
...
+#endif
     llvm_val[stmt] = builder->CreateLoad(
-        builder->CreateGEP(block_corner_coordinates,
-                           {tlctx->get_constant(0), tlctx->get_constant(0),
-                            tlctx->get_constant(stmt->index)}));
+#ifdef TI_LLVM_15
+        val_ty_as_array->getElementType(),
+#endif
+        builder->CreateGEP(
+#ifdef TI_LLVM_15
```
Comment (jim19930609) [inline taichi/codegen/llvm/codegen_llvm.cpp:2241]: ditto.

File: taichi/runtime/cpu/jit_cpu.cpp [hunk 6]
```diff
@@ -24,8 +26,15 @@
 #include "llvm/IR/Verifier.h"
 #include "llvm/IR/LLVMContext.h"
 #include "llvm/IR/LegacyPassManager.h"
+#ifdef TI_LLVM_15
+#include "llvm/MC/TargetRegistry.h"
+#else
 #include "llvm/Support/TargetRegistry.h"
+#endif
 #include "llvm/Support/DynamicLibrary.h"
+#ifdef TI_LLVM_15
```
Comment (jim19930609) [inline taichi/runtime/cpu/jit_cpu.cpp:35]: Let's group all these TI_LLVM_15 macros w.r.t header files.

File: taichi/runtime/cpu/jit_cpu.cpp [hunk 7]
```diff
@@ -286,7 +317,15 @@ std::unique_ptr<JITSession> create_llvm_jit_session_cpu(
     Arch arch) {
   TI_ASSERT(arch_is_cpu(arch));
   auto target_info = get_host_target_info();
-  return std::make_unique<JITSessionCPU>(tlctx, config, target_info.first,
+#ifdef TI_LLVM_15
+  auto EPC = SelfExecutorProcessControl::Create();
+  TI_ASSERT(EPC);
+#endif
+  return std::make_unique<JITSessionCPU>(tlctx,
+#ifdef TI_LLVM_15
```
Comment (jim19930609) [inline taichi/runtime/cpu/jit_cpu.cpp:325]: ditto

**Review by jim19930609 — APPROVED** [review]
Many thanks! 
In general it looks good to me, just some nit suggestions regarding code cleanliness.

**Comment by jim19930609** [comment]
Btw, would you mind setting up a windows-CI in github actions: https://github.com/taichi-dev/taichi/actions for TI_LLVM_15? Probably in a future PR? Thanks!

**Comment by python3kgae** [comment]
> Btw, would you mind setting up a windows-CI in github actions: https://github.com/taichi-dev/taichi/actions for TI_LLVM_15? Probably in a future PR? Thanks!

Sure. I'll try to add a windows-CI for llvm 15 which runs cpu only test.

**Review by jim19930609 — APPROVED** [review]
LGTM, Thanks!

**Pull request merged**